package com.jettyweb.redis;


import com.jettyweb.exception.SystemException;
import com.jettyweb.log.Log;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.exceptions.JedisConnectionException;

public class RedisTemplate {
	private Redis redis;

	public RedisTemplate(Redis redis) {
		this.redis = redis;
	}

	public <T> T execute(RedisCallBack<T> callback) {
		Jedis jedis = null;
		Exception e1 = null;
		for (int i = 0; i < redis.getTryCount(); i++) {
			try {
				e1 = null;
				jedis = redis.jedis();
				return callback.invoke(jedis);
			} catch (Exception e) {
				if (JedisConnectionException.class.isInstance(e)
						|| (e.getCause() != null && JedisConnectionException.class.isInstance(e.getCause()))) {
					Log.get("Redis.get#1").error("redis连接错误！" + e.getMessage(), e);
					if (jedis != null) {
						jedis.close();
						jedis = null;
					}
					e1 = e;
					continue;
				}
				Log.get("Redis.get#2").error("redis执行错误！" + e.getMessage(), e);
				if (jedis != null) {
					jedis.close();
					jedis = null;
				}
				SystemException.throwException(12342411, e.getMessage(), e);
			} finally {
				if (jedis != null) {
					jedis.close();
				}
			}
		}
		if (e1 != null) {
			throw new SystemException(12342422, e1.getMessage(), e1);
		}
		throw new SystemException(12342423, "未知redis异常");
	}
}
